File: frontend/src/types/calendar.ts
import { z } from "zod";

export const EventTypeSchema = z.enum([
  "earnings",
  "fomc",
  "fomc_minutes",
  "fed_speak",
  "economic",
  "opex",
  "quad_witching",
]);

export const ImportanceSchema = z.enum(["low", "medium", "high"]);

export const CalendarEventSchema = z.object({
  type: EventTypeSchema,
  title: z.string(),
  ticker: z.string().nullable().optional(),
  time: z.string().nullable().optional(),
  importance: ImportanceSchema,
});

export const CalendarDaySchema = z.object({
  date: z.string(),
  is_today: z.boolean(),
  events: z.array(CalendarEventSchema),
});

export const CalendarResponseSchema = z.object({
  days: z.array(CalendarDaySchema),
  notes: z.record(z.string()).default({}),
});

export type CalendarEvent = z.infer<typeof CalendarEventSchema>;
export type CalendarDay = z.infer<typeof CalendarDaySchema>;
export type CalendarResponse = z.infer<typeof CalendarResponseSchema>;
export type EventType = z.infer<typeof EventTypeSchema>;

File: backend/schemas/calendar.py
from typing import Literal

from pydantic import BaseModel, Field

EventType = Literal[
    "earnings", "fomc", "fomc_minutes", "fed_speak", "economic", "opex", "quad_witching"
]
Importance = Literal["low", "medium", "high"]


class CalendarEvent(BaseModel):
    type: EventType
    title: str
    ticker: str | None = None
    time: str | None = None
    importance: Importance = "medium"


class CalendarDay(BaseModel):
    date: str  # ISO YYYY-MM-DD
    is_today: bool
    events: list[CalendarEvent] = Field(default_factory=list)


class CalendarResponse(BaseModel):
    days: list[CalendarDay]
    # Per-section placeholder text. Surfaced in the UI as italic gray notes
    # — used today to flag the empty Fed-speakers list, like the watchlist's
    # "Sentiment unavailable" pattern.
    notes: dict[str, str] = Field(default_factory=dict)
